include/toolchain/gcc.h

```diff
@@ -94,6 +94,11 @@ do {                                                                    \
 #define __weak __attribute__((__weak__))
 #define __unused __attribute__((__unused__))
 
+/* Be *very* careful with this, you cannot filter out with -wno-deprecated,
+ * which has implications for -Werror
+ */
+#define __DEPRECATED_MACRO _Pragma("GCC warning \"Macro is deprecated\"")
```

Comment:
The suggested macro can even be used like so:

```C
#define __DEPRECATED_MACRO(cond)  __DEPRECATED_MACRO_##cond
#define __DEPRECATED_MACRO_0 _Pragma("GCC warning \"Macro is deprecated\"")
#define __DEPRECATED_MACRO_1 _Pragma("GCC error \"Gardening required on this macro\"")

#define __DEPRECATE_MACRO_ON(maj, min) \
   __DEPRECATED_MACRO(ZEPHYR_VERSION_CHECK(maj,min))
```

And we can then use `__DEPRECATE_MACRO_ON(1,11)`, ensuring a compile error will happen. (`ZEPHYR_VERSION_CHECK()` should evaluate to `0` or `1` given maj/min).